misc/python/materialize/cli/mzbench.py

```diff
@@ -0,0 +1,160 @@
+        "num_workers",
+        "seconds_taken",
+        "rows_per_second",
+        "grafana_url",
+    ]
+    results_writer = csv.DictWriter(sys.stdout, field_names)
+    results_writer.writeheader()
+
+    # We use check_output because check_call does not capture output
+    try:
+        subprocess.check_output(setup_benchmark, stderr=subprocess.STDOUT)
+    except (subprocess.CalledProcessError,) as e:
+        print(f"Setup benchmark failed! Output from failed command:\n{e.output}")
+        raise
+
+    for (worker_count, git_revision) in itertools.product(worker_counts, git_revisions):
+
+        child_env = os.environ.copy()
+        child_env["MZ_WORKERS"] = str(worker_count)
+        child_env["MZBUILD_WAIT_FOR_IMAGE"] = "true"
+        if git_revision:
+            child_env["MZBUILD_MATERIALIZED_TAG"] = f"unstable-{git_revision}"
+
+        try:
+            output = subprocess.check_output(
+                run_benchmark, env=child_env, stderr=subprocess.STDOUT
+            )
+        except (subprocess.CalledProcessError,) as e:
+            print(f"Setup benchmark failed! Output from failed command:\n{e.output}")
+            raise
+
+        # TODO: Replace parsing output from mzcompose with reading from a well known file or topic
+        for line in output.decode().splitlines():
+            if line.startswith("SUCCESS!"):
+                for token in line.split(" "):
+                    if token.startswith("seconds_taken="):
+                        seconds_taken = token[len("seconds_taken=") :]
+                    elif token.startswith("rows_per_sec="):
+                        rows_per_second = token[len("rows_per_sec=") :]
+            elif line.startswith("Grafana URL: "):
+                grafana_url = line[len("Grafana URL: ") :]
+
+        results_writer.writerow(
+            {
+                "git_revision": git_revision if git_revision else "NONE",
+                "num_workers": worker_count,
+                "seconds_taken": seconds_taken,
+                "rows_per_second": rows_per_second,
+                "grafana_url": grafana_url,
+            }
+        )
+
+
+def enumerate_cpu_counts() -> typing.List[int]:
+    """This program prints the number of CPU counts to benchmark on this machine.
+
+    We remove some percentage of CPU cores off the top for system / background processing. With
+    the CPUs that remain, we generate a list of evenly spaced worker counts. The list is limited
+    by the number of trials desired. This is meant to help us explore the number of CPUs that
+    should be dedicated to MZ_WORKERS, not as a prescription for the correct values to choose.
+
+    On a Macbook with 8 cores, this will return [3, 2, 1].
+
+    On a 56 core machine, this returns [24, 18, 12, 6].
+
+    On a 96 core machine, this returns [41, 30, 20, 10].
+    """
+
+    # 15% overhead, divide by 2 because we assume half are hyperthreads
+    max_cpus = round(psutil.cpu_count(logical=False) * 0.85)
```

Comment:
the comment above mentions that you divide by two but its not clear to me where thats applied

Reply:
This comment needs to be updated since we are now using `psutil` to get the number of phsyical cores. The previous version had a divide by two because it was using `os.cpu_count()`, which includes logical cores.